src/webui/api/appcontroller.cpp

```diff
@@ -1312,6 +1357,89 @@ void AppController::setCookiesAction()
     setResult(QString());
 }
 
+void AppController::clientDataAction()
+{
+    const QString keysParam {params()[u"keys"_s]};
+    if (keysParam.isEmpty())
+    {
+        setResult(m_clientData);
+        return;
+    }
+
+    QJsonParseError jsonError;
+    const auto keysJsonDocument = QJsonDocument::fromJson(keysParam.toUtf8(), &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+        throw APIError(APIErrorType::BadParams, jsonError.errorString());
+    if (!keysJsonDocument.isArray())
+        throw APIError(APIErrorType::BadParams, tr("keys must be an array"));
+
+    QJsonObject clientData;
+    for (const QJsonValue &keysJsonVal : asConst(keysJsonDocument.array()))
+    {
+        if (!keysJsonVal.isString())
+            throw APIError(APIErrorType::BadParams, tr("key must be a string"));
+
+        const QString &key = keysJsonVal.toString();
+        if (const auto iter = m_clientData.constFind(key); iter != m_clientData.constEnd())
+            clientData.insert(key, iter.value());
+    }
+
+    setResult(clientData);
+}
+
+void AppController::setClientDataAction()
+{
+    requireParams({u"data"_s});
+    QJsonParseError jsonError;
+    const auto dataJsonDocument = QJsonDocument::fromJson(params()[u"data"_s].toUtf8(), &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+        throw APIError(APIErrorType::BadParams, jsonError.errorString());
+    if (!dataJsonDocument.isObject())
+        throw APIError(APIErrorType::BadParams, tr("data must be an object"));
+
+    QJsonObject clientData = m_clientData;
+    bool dataModified = false;
+    const QJsonObject dataJsonObject = dataJsonDocument.object();
+    for (auto it = dataJsonObject.constBegin(), end = dataJsonObject.constEnd(); it != end; ++it)
+    {
+        const QString &key = it.key();
+        const QJsonValue &value = it.value();
+
+        if (value.isNull())
```

Comment:
BTW, having explicit `removeData` action provided would allow you to store `null` values. Maybe it would be useful to someone.

Reply:
KISS: I can't see any use case for storing `null` right now. I think it's better to have this interface for now and we can change down the line if a new use case emerges.